src/ModelContextProtocol.Analyzers/XmlToDescriptionGenerator.cs

```diff
@@ -24,27 +25,38 @@ public sealed class XmlToDescriptionGenerator : IIncrementalGenerator
 
     public void Initialize(IncrementalGeneratorInitializationContext context)
     {
-        // Use ForAttributeWithMetadataName for each MCP attribute type
-        var toolMethods = CreateProviderForAttribute(context, McpServerToolAttributeName);
-        var promptMethods = CreateProviderForAttribute(context, McpServerPromptAttributeName);
-        var resourceMethods = CreateProviderForAttribute(context, McpServerResourceAttributeName);
-
-        // Combine all three providers
-        var allMethods = toolMethods
-            .Collect()
-            .Combine(promptMethods.Collect())
-            .Combine(resourceMethods.Collect())
+        // Extract method information for all MCP tools, prompts, and resources.
+        var allMethods = CreateProviderForAttribute(context, McpServerToolAttributeName).Collect()
+            .Combine(CreateProviderForAttribute(context, McpServerPromptAttributeName).Collect())
+            .Combine(CreateProviderForAttribute(context, McpServerResourceAttributeName).Collect())
             .Select(static (tuple, _) =>
             {
-                var ((tool, prompt), resource) = tuple;
-                return tool.AddRange(prompt).AddRange(resource);
+                var ((tools, prompts), resources) = tuple;
+                return tools.AddRange(prompts).AddRange(resources);
             });
 
-        // Combine with compilation to get well-known type symbols.
-        var compilationAndMethods = context.CompilationProvider.Combine(allMethods);
+        // Report diagnostics.
+        context.RegisterSourceOutput(allMethods, static (spc, methods) =>
```

Comment:
This separation of diagnostic reporting into a separate callback is an interesting approach. I think it might provide a viable workaround for https://github.com/dotnet/runtime/issues/92509. Curious if you've applied this approach elsewhere before, or if there are any drawbacks to it.

Reply:
If there are any diagnostics, I expect that equality will be completely broken here. Our recommendation is to have separate analyzers for diagnostics and not do any reporting in the generator.